providers/signatory-ledger-cosval/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+//! Ed25519 provider for the ledger cosmos validator app
+
+#![crate_name = "signatory_ledger_cosval"]
+#![crate_type = "lib"]
+#![deny(warnings, missing_docs, trivial_casts, trivial_numeric_casts)]
+#![deny(unsafe_code, unused_import_braces, unused_qualifications)]
+#![doc(
+html_logo_url = "https://raw.githubusercontent.com/tendermint/signatory/master/img/signatory-rustacean.png",
+html_root_url = "https://docs.rs/signatory-ledger-cosval/0.0.1"
+)]
+
+extern crate signatory;
+extern crate ledger_cosmos;
+
+use std::sync::Mutex;
+use std::sync::Arc;
+
+use signatory::{
+    ed25519::Ed25519Signature,
+    ed25519::Ed25519PublicKey,
+    error::{Error, ErrorKind},
+    PublicKeyed,
+    Signer,
+};
+
+/// ed25519 signature provider for the ledger cosmos validator app
+#[allow(dead_code)]
+pub struct Ed25519CosmosAppSigner {
+    app: Arc<Mutex<ledger_cosmos::CosmosValidatorApp>>
+}
+
+impl Ed25519CosmosAppSigner {
+    /// Create a new ed25519 signer based on ledger nano s - cosmos validator app
+    pub fn connect() -> Result<Self, Error> {
+        // TODO: Maybe use this to pass other derivation path
+
+        match ledger_cosmos::CosmosValidatorApp::connect() {
+            Ok(validator_app) => {
+                let app = Arc::new(Mutex::new(validator_app));
+                Ok(Ed25519CosmosAppSigner { app })
+            },
+            Err(err) => Err(
+                Error::new(ErrorKind::ProviderError, Some(&err.to_string()))
+            )
+        }
+    }
+}
+
+impl PublicKeyed<Ed25519PublicKey> for Ed25519CosmosAppSigner {
+    /// Returns the public key that corresponds cosmos validator app connected to this signer
+    fn public_key(&self) -> Result<Ed25519PublicKey, Error> {
+        let app = self.app.lock().unwrap();
+
+        match app.public_key() {
+            Ok(pk) => Ok(Ed25519PublicKey(pk)),
+            Err(err) => Err(
+                Error::new(ErrorKind::ProviderError, Some(&err.to_string()))
+            )
+        }
+    }
+}
+
+impl Signer<Vec<u8>, Ed25519Signature> for Ed25519CosmosAppSigner {
```

Comment:
Would suggest `impl<'a> Signer<&'a [u8], ...>` here, as the `ByteSigner` trait is gated on that, and that's what the KMS trait object is gated on. See:

- https://github.com/tendermint/signatory/blob/master/src/signer/bytes.rs#L14
- https://github.com/tendermint/kms/blob/master/src/ed25519/signer/mod.rs#L22

Reply:
no problem, I will change that